### PR TITLE
Drop support for legacy Python 2.7

### PR DESCRIPTION
**Pull request by hugovk**
For #760

### CODE REVIEW

File: markdown/util.py [hunk 1]
```diff
@@ -20,32 +19,15 @@
...
-=============================================================================
-"""
-PY3 = sys.version_info[0] == 3
 PY37 = (3, 7) <= sys.version_info
 
-if PY3:  # pragma: no cover
-    string_type = str
-    text_type = str
-    int2str = chr
-    iterrange = range
```
Comment (waylan) [inline markdown/util.py:42]: It is entirely possible that third party extensions are importing these variables. Perhaps we should leave them defined, even though we don't use them. We might even consider wrapping them with `deprecated` and issue a DeprecationWarning on use. Of course, in a future release they would then be removed.

File: markdown/__init__.py [hunk 2]
```diff
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (waylan) [inline markdown/__init__.py:1]: It appears that this was removed from all files. Why? This is not related to Python version at all and is still a valid comment to have for people who use certain editors.

File: markdown/core.py [hunk 3]
```diff
@@ -321,15 +318,15 @@ def convertFile(self, input=None, output=None, encoding=None):
 
         # Read the source
         if input:
-            if isinstance(input, util.string_type):
+            if isinstance(input, str):
                 input_file = codecs.open(input, mode="r", encoding=encoding)
```
Comment (waylan) [inline markdown/core.py:322]: This code makes heavy use of `codecs.open` because it was developed prior to the `io` lib existing. It should probably be reworked to use the default `open` builtin. That said, it might make sense to do that as a separate PR.

**Review by waylan — CHANGES_REQUESTED** [review]
Wow. That was a lot of work. Thanks. I took a rather quick look at the changes and have noted a few minor issues below. Get those addressed and I'll take a closer look when I have more time.

File: markdown/__init__.py (hunk 2 above)
Comment (waylan) [reply markdown/__init__.py:1]: So I took another look at this and noticed that this change was part of commit c62d978, which was generated by `pyupgrade`.  The [documentation][1] for `pyupgrade` points to [PEP3120][2] as justification for removing the comments. I had mis-remembered why we needed those comments (I was incorrectly thinking the comment was for editors, not the Python compiler). Given the above, I am fine with leaving the change as-is. 

[1]: https://github.com/asottile/pyupgrade#-coding--comment
[2]: https://www.python.org/dev/peps/pep-3120/

File: tests/test_apis.py [hunk 4]
```diff
@@ -442,12 +434,6 @@ def tearDown(self):
         # Reset warning behavior back to default
         warnings.simplefilter('default')
 
-    def testNonUnicodeSource(self):
-        """ Test falure on non-unicode source text. """
-        if not PY3:
-            source = "foo".encode('utf-16')
-            self.assertRaises(UnicodeDecodeError, markdown.markdown, source)
-
```
Comment (waylan) [inline tests/test_apis.py]: Let's leave this test in. It is entirely reasonable to expect that a user could pass in something other than `utf-8` source text and we should continue to return an appropriate error. I haven't checked this, but if the test was removed because it was failing, then we need to make some changes to the code to make it pass again.

**Comment by waylan** [comment]
I did a search for `TODO` comments in the code and found one comment relating to dropping PY2 at [markdown/test_tools.py#L150-L156][1]:

```python
# Define LegacyTestCase class with metaclass in Py2 & Py3 compatible way.
# See https://stackoverflow.com/a/38668373/866026
# TODO: If/when py2 support is dropped change to:
# class LegacyTestCase(unittest.Testcase, metaclass=LegacyTestMeta)


class LegacyTestCase(LegacyTestMeta('LegacyTestCase', (unittest.TestCase,), {'__slots__': ()})):
```

We should make the change anticipated in that comment (and remove the comment) as part of this PR.

It does not appear that any other `TODO` comments are related to dropping PY2. However, if anyone finds any, feel free to point them out.

[1]: https://github.com/Python-Markdown/markdown/blob/dbdb41469e4101c0fdf4b30f69be35d9cd801430/markdown/test_tools.py#L150-L156

**Comment by waylan** [comment]
Also, a comment needs to be added to the release notes.

File: markdown/util.py (hunk 1 above)
Comment (hugovk) [reply markdown/util.py:42]: I've reverted them in e51e15b3f518ac62d1e99223eb58a5c6b65ebae5 and added a deprecation comment for now.

File: markdown/core.py (hunk 3 above)
Comment (hugovk) [reply markdown/core.py:322]: Yes, I think a separate PR would be sensible.

**Comment by hugovk** [comment]
Thanks, updated!

**Review by waylan — APPROVED** [review]

**Comment by waylan** [comment]
Not sure why the latest [Travis build][1] isn't showing up, but it is passing.

@hugovk note the changes I made in cfa1aa6 which marks the variables as deprecated. Conveniently, we already have a vendored backport of `pep562` included, so it was easy to make use of it.

[1]: https://travis-ci.org/Python-Markdown/markdown/builds/602322487?utm_medium=notification&utm_source=github_status

**Comment by hugovk** [comment]
Great, thanks for the finishing touches!
